File: Python/Horner.py
#O(n)
def horner_iterative(coeffs: list, x_value: float) -> float:
    result = coeffs[0]
    for coefficient in coeffs[1:]:
        result = result * x_value + coefficient
    return result

print(horner_iterative([3, 3, -2, 11], 2.5))

#O(n)
def horner_recursive(coeffs: list, x: float, degree: int = None) -> float:
    if degree is None:
        degree = len(coeffs) - 1
    if degree == 0:
        return coeffs[0]
    return x * horner_recursive(coeffs, x, degree - 1) + coeffs[degree]

print(horner_recursive([3, 3, -2, 11], 2.5))

File: CPP/Horner.cpp
#include <iostream>

using namespace std;

double hornerIterative(double coeffs[], int degree, double x) {
    double result = coeffs[0];
    for (int i = 1; i <= degree; i++) {
        result = result * x + coeffs[i];
    }
    return result;
}

double hornerRecursive(double coeffs[], int degree, double x) {
    if (degree == 0) {
        return coeffs[0];
    }
    return x * hornerRecursive(coeffs, degree - 1, x) + coeffs[degree];
}

int main() {
    double hor[] = {3, 3, -2, 11};
    cout << hornerIterative(hor, 3, 2.5) << endl;
    cout << hornerRecursive(hor, 3, 2.5) << endl;
}